design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+apiVersion: compute.gcp.crossplane.io/v1alpha3
+kind: Subnetwork
+metadata:
+  name: coolsubnet
+spec:
+  forProvider:
+    network: /projects/foo/global/networks/bar
+    networkRef:
+      name: foobar
+    networkSelector:
+      matchConfiguration: true
+      matchLabels:
+        example: label
+```
+
+The above example shows a GCP `Subnetwork` managed resource that uses a
+reference selector. `networkSelector` sets the `networkRef` which in turn sets
+the value of `network`. The reference selector contains two optional fields:
+
+* `matchConfiguration`: Match only managed resources that are part of the same
+  `ResourceConfiguration` (i.e. have the same `configurationRef`) as the
+  selecting resource.
+* `matchLabels`: Match only managed resources with the supplied labels.
+
+If a reference field is set, its corresponding selector field is ignored. If the
+selector field is unset, it is ignored. If the specified selector matches
+multiple managed resources one is chosen at random. Note however that setting
+both `matchConfiguration` and `matchLabels` can guarantee that at most one
+provisioned managed resource will match the selector.
+
+### External Names
+
+Managed resources have a name and an _external name_. The former identifies the
+managed resource in the Kubernetes API, while the latter identifies the resource
+in the external system. Managed resources that have reached `v1beta1` allow both
+managed resource and resource claim authors to [control the name of their
+external resource] using the `crossplane.io/external-name` annotation:
+
+* The name of the external resource is set to the value of the annotation.
+* If the annotation is absent it is set to the managed resource's name.
+* The annotation is propagated from a resource claim to any managed resource it
+  dynamically provisions.
+
+This final point may pose a problem when one resource claim may provision many
+managed resources. If a resource claim set the `crossplane.io/external-name`
+annotation and referenced a `ResourceConfigurationClass` that specified two
+managed resources of the same kind, both dynamically provisioned resources would
+attempt to use the same external name. External names are unique at different
+scopes (global, project, region) for different resources so this will not be an
+issue all resources, but it may for some.
+
+This document proposes two new external naming annotations be introduced to help
+configuration class authors avoid external name conflicts:
+
+* `crossplane.io/external-name-prefix` - A value prepended to the external name
+  of a managed resource.
+* `crossplane.io/external-name-suffix` - A value appended to the external name
+  of a managed resource.
+
+These annotations would be set on the provisioned managed resources and parsed
+by their controllers alongside the existing `crossplane-io/external-name`
+annotation, resulting in an actual external name of the form
+`{external-name-prefix}{external-name}{external-name-suffix}`.
```

Comment:
Does this mean if all three annotations are set that conflicts could still occur?

Reply:
Only if all three were set to the same values - if the prefix or suffix differed for each resource there would be no conflict.